File: packages/extension/src/core/pointer.ts
/**
 * Read Duet pointer file (~/.org.ve68.duet).
 *
 * Pointer file is a minimal JSON that tells all Duet components
 * where to find DuetData (local cache) and DuetConfig (cloud source of truth).
 *
 * For tests: env DUET_POINTER_FILE overrides the default path.
 */
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

export interface Pointer {
    machine: string;
    duetDataPath: string;
    duetConfigPath: string;
}

/**
 * Get path to pointer file.
 * Uses DUET_POINTER_FILE env for testing, otherwise ~/.org.ve68.duet
 */
export function getPointerPath(): string {
    return process.env.DUET_POINTER_FILE || path.join(os.homedir(), '.org.ve68.duet');
}

/**
 * Read and validate pointer file.
 * Returns null if file not found, unreadable, or missing required fields.
 */
export function readPointer(): Pointer | null {
    const pointerPath = getPointerPath();
    try {
        const content = fs.readFileSync(pointerPath, 'utf-8');
        const data = JSON.parse(content);
        if (
            typeof data.machine === 'string' &&
            typeof data.duetDataPath === 'string' &&
            typeof data.duetConfigPath === 'string'
        ) {
            return data as Pointer;
        }
        return null;
    } catch {
        return null;
    }
}

/**
 * Read {machine}.json from DuetConfig.
 * Returns null if pointer or machine config is missing/unreadable.
 */
export function readMachineConfig(): Record<string, unknown> | null {
    const pointer = readPointer();
    if (!pointer) {
        return null;
    }
    const configPath = path.join(pointer.duetConfigPath, `${pointer.machine}.json`);
    try {
        const content = fs.readFileSync(configPath, 'utf-8');
        return JSON.parse(content);
    } catch {
        return null;
    }
}

/**
 * Read port from machine config.
 * Returns 19680 (default) if config is missing or port is not set.
 */
export function readPort(): number {
    const config = readMachineConfig();
    const port = config?.port;
    return typeof port === 'number' ? port : 19680;
}

File: packages/backend/pointer.py
"""Pointer file reader for Duet backend.

The pointer file (~/.org.ve68.duet) is a JSON file that points to:
- DuetData: local cache directory
- DuetConfig: cloud-synced configuration directory
- machine: identifier for machine-specific config

Pointer file format:
{
    "machine": "mac_work",
    "duetDataPath": "/Users/.../DuetData",
    "duetConfigPath": "/Users/.../DuetConfig"
}

For testing: set env DUET_POINTER_FILE to override default path.
"""

import json
import os
from pathlib import Path
from typing import TypedDict


class PointerData(TypedDict):
    """Contents of the pointer file."""

    machine: str
    duetDataPath: str
    duetConfigPath: str


class PointerNotFoundError(Exception):
    """Raised when pointer file does not exist."""

    pass


class PointerInvalidError(Exception):
    """Raised when pointer file is invalid or missing required fields."""

    pass


def get_pointer_path() -> Path:
    """Get path to pointer file.

    Default: ~/.org.ve68.duet
    Override: env DUET_POINTER_FILE (for testing)
    """
    env_path = os.environ.get("DUET_POINTER_FILE")
    if env_path:
        return Path(env_path)
    return Path.home() / ".org.ve68.duet"


def read_pointer() -> PointerData:
    """Read and validate pointer file.

    Returns:
        PointerData with machine, duetDataPath, duetConfigPath.

    Raises:
        PointerNotFoundError: If pointer file does not exist.
        PointerInvalidError: If file is invalid JSON or missing required fields.
    """
    pointer_path = get_pointer_path()

    if not pointer_path.exists():
        raise PointerNotFoundError(
            f"Pointer file not found: {pointer_path}. "
            "Run Duet Host to create it."
        )

    try:
        data = json.loads(pointer_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise PointerInvalidError(f"Invalid JSON in pointer file: {e}") from e

    # Validate required fields
    required = ["machine", "duetDataPath", "duetConfigPath"]
    missing = [f for f in required if f not in data or not data[f]]
    if missing:
        raise PointerInvalidError(
            f"Pointer file missing required fields: {', '.join(missing)}"
        )

    # Validate types
    for field in required:
        if not isinstance(data[field], str):
            raise PointerInvalidError(
                f"Pointer field '{field}' must be a string, got {type(data[field]).__name__}"
            )

    return PointerData(
        machine=data["machine"],
        duetDataPath=data["duetDataPath"],
        duetConfigPath=data["duetConfigPath"],
    )


def pointer_exists() -> bool:
    """Check if pointer file exists (without reading it)."""
    return get_pointer_path().exists()
